trimesh/util.py

```diff
@@ -1167,6 +1172,102 @@ def array_to_string(array,
 
     return formatted
 
+def structured_array_to_string(array,
+                    col_delim=' ',
+                    row_delim='\n',
+                    digits=8,
+                    value_format='{}'):
+    """
+    Convert an unstructured array into a string with a specified
+    number of digits and delimiter. The reason thisexists is
+    that the basic numpy array to string conversionsare
+    surprisingly bad.
+
+    Parameters
+    ------------
+    array : (n,) or (n, d) float or int
+       Data to be converted
+       If shape is (n,) only column delimiter will be used
+    col_delim : str
+      What string should separate values in a column
+    row_delim : str
+      What string should separate values in a row
+    digits : int
+      How many digits should floating point numbers include
+    value_format : str
+       Format string for each value or sequence of values
+       If multiple values per value_format it must divide
+       into array evenly.
+
+    Returns
+    ----------
+    formatted : str
+       String representation of original array
+    """
+    # convert inputs to correct types
+    array = np.asanyarray(array)
+    digits = int(digits)
+    row_delim = str(row_delim)
+    col_delim = str(col_delim)
+    value_format = str(value_format)
+
+    # abort for non-flat arrays
+    if len(array.shape) > 1:
+        raise ValueError('conversion only works on 1D/2D arrays not %s!',
+                         str(array.shape))
+
+    # abort for unstructured arrays
+    if array.dtype.names is None:
+        raise ValueError('array is not structured, use array_to_string instead')
+
+
+    # do not allow a value to be repeated in a value format
+    if value_format.count('{') > 1:
+        raise ValueError(
+            'value_format %s is invalid, repeating unstructured array '
+            + 'values is unsupported', value_format)
+
+    format_str = ''
+    for name in array.dtype.names:
+        kind = array[name].dtype.kind
+        element_row_length = (
+            array[name].shape[1] if len(array[name].shape) == 2 else 1)
+        if kind in ['i', 'u']:
+            # integer types don't need a specified precision
+            element_format_str = (value_format + col_delim)
+        elif kind == 'f':
+            # add the digits formatting to floats
+            element_format_str = value_format.replace(
+                '{}', '{:.' + str(digits) + 'f}') + col_delim
+        else:
+            raise ValueError('dtype %s not convertible!',
+                            array.dtype)
+        format_str += element_row_length * element_format_str
+
+    # length of extra delimiters at the end
+    format_str = format_str[:-len(col_delim)] + row_delim
+
+    # expand format string to whole array
+    format_str *= len(array)
+
+    # Flatten the array to a list as numpy array column type cannot be mixed
+    partially_flattened = [
```

Comment:
I'm not super happy with this part and would have preferred to keep this as a numpy reshape but that doesn't work because numpy doesn't support mixed types in an array column.

Maybe a custom generator would be more efficient than list comprehension + concatenation, but I figured this works for now and that is an optimization that can be done if required.

Reply:
Yeah the flattening operation is tough. I was playing with this a bit, if you stay in numpy-land and only loop through field names it's quite a bit faster (~45x):
`flattened = np.hstack([array[k].reshape((count, -1)) for k in array.dtype.names]).reshape(-1)`

As you say it upgrades everything to a float. On the other hand if we wangled `value_format= '{:0.0f}'` or possibly `value_format.replace('{}', '{:0.0f}')` for those fields we might be able to cut off the decimal using the string format we're doing anyway.